### PR TITLE
Adding Binary Sort Algorithm in JavaScript

### PR DESCRIPTION
**Pull request by vale-c**
**Hi!**  
Thanks for considering contributing to this ever-growing list of algorithm and data structure implementations.
Your contribution is valuable. Use the follow template for creating Pull Requests
---

**Fixes [#435](https://github.com/FOSS-UCSC/FOSSALGO/issues/435) **

By submitting this pull request I confirm I've read and complied with the below declarations.

-   [X] I have read the [Contribution guidelines](https://github.com/FOSS-UCSC/FOSSALGO/blob/master/CONTRIBUTING.md) 
and I am confident that my PR reflects them.

-   [X] I have followed the [coding guidelines](https://github.com/FOSS-UCSC/FOSSALGO/blob/master/CONTRIBUTING.md#cs) 
for this project.

-   [X] My code follows the [skeleton code structure](https://github.com/FOSS-UCSC/FOSSALGO/blob/master/CONTRIBUTING.md#sample).

-   [X] This pull request has a descriptive title. 
For example, `Added {Algorithm/DS name} [{Language}]`, not `Update README.md` or `Added new code`.

-   [X] This pull request shall only be merged once the build passes.

-   [X] I have mentioned the issue number correctly (with hyper-link) in this pull request description.

### CODE REVIEW

File: algorithms/ar-binsrt/javascript/binarySort.js [hunk 1]
```diff
@@ -0,0 +1,45 @@
+function binarySort(item, sortedList, low = 0, high = sortedList.length - 1) {
+  if (low == high) {
+    // hit end of sortedList - done
+    return item > sortedList[low] ? low + 1 : low
```
Comment (RavinduSachintha) [inline algorithms/ar-binsrt/javascript/binarySort.js:4]: Codacy found an issue: [Missing semicolon.](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6318101)

File: algorithms/ar-binsrt/javascript/binarySort.js [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+  }
+
+  // get midpoint of list and item value
+  let mid = low + Math.floor((high - low) / 2),
+    itemCompare = sortedList[mid]
```
Comment (RavinduSachintha) [inline algorithms/ar-binsrt/javascript/binarySort.js:9]: Codacy found an issue: [Missing semicolon.](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6318101)

File: algorithms/ar-binsrt/javascript/binarySort.js [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+}
+
+// insert values
+let startTime = Date.now()
+
+for (let i = 1; i < 1000; i++) {
+  // generate a rand num and insert
+  let nextValue = Math.floor(Math.random() * 2000000)
+
+  resultIndexList.splice(binarySort(nextValue, resultIndexList), 0, nextValue)
```
Comment (RavinduSachintha) [inline algorithms/ar-binsrt/javascript/binarySort.js:42]: Codacy found an issue: [Missing semicolon.](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6318101)

File: algorithms/ar-binsrt/javascript/binarySort.js [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+}
+
+// seed list of random ascending (sorted) numbers
+let lastValue = 1,
+  resultIndexList = [lastValue]
+
+for (let i = 1; i < 10000; i++) {
+  // generate a random higher number than prev one
+  lastValue = Math.floor(Math.random() * 500) + lastValue
+  resultIndexList.push(lastValue)
```
Comment (RavinduSachintha) [inline algorithms/ar-binsrt/javascript/binarySort.js:32]: Codacy found an issue: [Missing semicolon.](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6318101)

File: algorithms/ar-binsrt/javascript/binarySort.js [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    return item > sortedList[low] ? low + 1 : low
+  }
+
+  // get midpoint of list and item value
+  let mid = low + Math.floor((high - low) / 2),
+    itemCompare = sortedList[mid]
+
+  if (item > itemCompare) {
+    // work higher end of list
+    return binarySort(item, sortedList, mid + 1, high)
```
Comment (RavinduSachintha) [inline algorithms/ar-binsrt/javascript/binarySort.js:13]: Codacy found an issue: [Missing semicolon.](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6318101)

File: algorithms/ar-binsrt/javascript/binarySort.js [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+    return binarySort(item, sortedList, low, mid)
+  }
+
+  // found equal value - done
+  return mid
+}
+
+// seed list of random ascending (sorted) numbers
+let lastValue = 1,
+  resultIndexList = [lastValue]
```
Comment (RavinduSachintha) [inline algorithms/ar-binsrt/javascript/binarySort.js:27]: Codacy found an issue: [Missing semicolon.](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6318101)

File: algorithms/ar-binsrt/javascript/binarySort.js [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+let startTime = Date.now()
+
+for (let i = 1; i < 1000; i++) {
+  // generate a rand num and insert
+  let nextValue = Math.floor(Math.random() * 2000000)
+
+  resultIndexList.splice(binarySort(nextValue, resultIndexList), 0, nextValue)
+}
+
+console.log(`Duration: ${Date.now() - startTime}ms`)
```
Comment (RavinduSachintha) [inline algorithms/ar-binsrt/javascript/binarySort.js:45]: Codacy found an issue: [Missing semicolon.](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6318101)

File: algorithms/ar-binsrt/javascript/binarySort.js [hunk 8]
```diff
@@ -0,0 +1,45 @@
+function binarySort(item, sortedList, low = 0, high = sortedList.length - 1) {
+  if (low == high) {
```
Comment (RavinduSachintha) [inline algorithms/ar-binsrt/javascript/binarySort.js:2]: Codacy found an issue: [Expected '===' and instead saw '=='.](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6318101)

File: algorithms/ar-binsrt/javascript/binarySort.js (hunk 7 above)
Comment (RavinduSachintha) [inline algorithms/ar-binsrt/javascript/binarySort.js:45]: Codacy found an issue: [Unexpected console statement.](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6318101)

File: algorithms/ar-binsrt/javascript/binarySort.js [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+    return binarySort(item, sortedList, mid + 1, high)
+  }
+
+  if (item < itemCompare) {
+    // work lower end of list
+    return binarySort(item, sortedList, low, mid)
+  }
+
+  // found equal value - done
+  return mid
```
Comment (RavinduSachintha) [inline algorithms/ar-binsrt/javascript/binarySort.js:22]: Codacy found an issue: [Missing semicolon.](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6318101)

File: algorithms/ar-binsrt/javascript/binarySort.js [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+  resultIndexList = [lastValue]
+
+for (let i = 1; i < 10000; i++) {
+  // generate a random higher number than prev one
+  lastValue = Math.floor(Math.random() * 500) + lastValue
+  resultIndexList.push(lastValue)
+}
+
+// insert values
+let startTime = Date.now()
```
Comment (RavinduSachintha) [inline algorithms/ar-binsrt/javascript/binarySort.js:36]: Codacy found an issue: [Missing semicolon.](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6318101)

File: algorithms/ar-binsrt/javascript/binarySort.js [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+
+for (let i = 1; i < 1000; i++) {
+  // generate a rand num and insert
+  let nextValue = Math.floor(Math.random() * 2000000)
```
Comment (RavinduSachintha) [inline algorithms/ar-binsrt/javascript/binarySort.js:40]: Codacy found an issue: [Missing semicolon.](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6318101)

File: algorithms/ar-binsrt/javascript/binarySort.js [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+    itemCompare = sortedList[mid]
+
+  if (item > itemCompare) {
+    // work higher end of list
+    return binarySort(item, sortedList, mid + 1, high)
+  }
+
+  if (item < itemCompare) {
+    // work lower end of list
+    return binarySort(item, sortedList, low, mid)
```
Comment (RavinduSachintha) [inline algorithms/ar-binsrt/javascript/binarySort.js:18]: Codacy found an issue: [Missing semicolon.](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6318101)

File: algorithms/ar-binsrt/javascript/binarySort.js [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+  return mid
+}
+
+// seed list of random ascending (sorted) numbers
+let lastValue = 1,
+  resultIndexList = [lastValue]
+
+for (let i = 1; i < 10000; i++) {
+  // generate a random higher number than prev one
+  lastValue = Math.floor(Math.random() * 500) + lastValue
```
Comment (RavinduSachintha) [inline algorithms/ar-binsrt/javascript/binarySort.js:31]: Codacy found an issue: [Missing semicolon.](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6318101)

File: algorithms/ar-binsrt/javascript/binarySort.js [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+let startTime = Date.now();
+
+for (let i = 1; i < 1000; i++) {
+  // generate a rand num and insert
+  let nextValue = Math.floor(Math.random() * 2000000);
+
+  resultIndexList.splice(binarySort(nextValue, resultIndexList), 0, nextValue);
+}
+
+console.log(`Duration: ${Date.now() - startTime}ms`);
```
Comment (RavinduSachintha) [inline algorithms/ar-binsrt/javascript/binarySort.js:45]: Codacy found an issue: [Unexpected console statement.](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6318101)

**Comment by Danushka96** [comment]
Hi @vale-c 
Remove the console output and rename the file to lowercase, if needed use underscore

**Review by RavinduSachintha — APPROVED** [review]

**Comment by RavinduSachintha** [comment]
@vale-c can you add your contribution to the algorithm chart in the main readme file. don't change anything else. And also push the changes to this PR.

**Review by RavinduSachintha — APPROVED** [review]

**Comment by RavinduSachintha** [comment]
@all-contributors please add @vale-c for code and doc

**Comment by RavinduSachintha** [comment]
@vale-c thank you for your contribution. keep it up :+1: and make sure to star this repo :wink:

**Pull request merged**